src/cdk-experimental/table/virtual-rows.ts

```diff
@@ -0,0 +1,292 @@
+      @Inject(VIRTUAL_SCROLL_STRATEGY) protected readonly _scrollStrategy: VirtualScrollStrategy,
+      @SkipSelf() private readonly _viewport: CdkVirtualScrollViewport) {
+    positioningListener.setListener(this);
+
+    // Force the table to enable `fixedLayout` to prevent column widths from changing as the user
+    // scrolls. This also enables caching in the table's sticky styler which reduces calls to
+    // expensive DOM APIs, such as `getBoundingClientRect()`, and improves overall performance.
+    this._table.fixedLayout = true;
+
+    // Update sticky styles for header rows when either the render range or sticky state change.
+    combineLatest([this._renderedContentOffsetChanged, this._headerRowStickyUpdates])
+      .pipe(takeUntil(this._destroyed))
+      .subscribe(([offset, update]) => {
+        this.stickHeaderRows(offset, update);
+      });
+
+    // Update sticky styles for footer rows when either the render range or sticky state change.
+    combineLatest([this._renderedContentOffsetChanged, this._footerRowStickyUpdates])
+      .pipe(takeUntil(this._destroyed))
+      .subscribe(([offset, update]) => {
+        this.stickFooterRows(offset, update);
+      });
+
+    // Update viewChange subscribers when the virtual scroll viewport's rendered range changes.
+    this._viewport.renderedRangeStream.pipe(takeUntil(this._destroyed)).subscribe(range => {
+      this._renderedRange = range;
+      this._table.viewChange.next(range);
+    });
+
+    this._viewport.attach(this);
+  }
+
+  ngOnDestroy() {
+    this._destroyed.next();
+    this._destroyed.complete();
+  }
+
+  measureRangeSize(range: ListRange, orientation: 'horizontal' | 'vertical'): number {
+    return 0;
+  }
+
+  stickyColumnsUpdated(update: StickyUpdate): void {
+    // no-op
+  }
+
+  stickyEndColumnsUpdated(update: StickyUpdate): void {
+    // no-op
+  }
+
+  stickyHeaderRowsUpdated(update: StickyUpdate): void {
+    this._headerRowStickyUpdates.next(update);
+  }
+
+  stickyFooterRowsUpdated(update: StickyUpdate): void {
+    this._footerRowStickyUpdates.next(update);
+  }
+
+  private getDataStream(dataSource: CdkTableDataSourceInput<T>): Observable<T[] | ReadonlyArray<T>> {
+    if (isDataSource(dataSource)) {
+      return dataSource.connect(this._table);
+    } else if (isObservable(dataSource)) {
+      return dataSource;
+    }
+    return observableOf(dataSource);
+  }
+
+  private stickHeaderRows(offsetFromTop: number, update: StickyUpdate) {
+    if (!update.sizes || !update.elements) {
+      return;
+    }
+
+    // FIXME DO NOT SUBMIT: sum could be incorrect if some header rows are not sticky. Consider
+    //  adding `offsets` property to `StickyUpdate` that indicates the offset of each row from the
+    //  bottom/top, so the offset doesn't have to be computed manually.
+    let sumOfRowHeights = 0;
+    for (let i = 0; i < update.elements.length; i++) {
+      sumOfRowHeights += i > 0 ? update.sizes[i]! : 0;
```

Comment:
Nit: it's one less operation per row to do

if (i > 0) {
  sumOfRowHeights =+ update.sizes[i];
}